src/engine/MILPEncoder.cpp

```diff
@@ -126,3 +129,85 @@ void MILPEncoder::encodeReLUConstraint( GurobiWrapper &gurobi, ReluConstraint
     terms.append( GurobiWrapper::Term( -sourceUb, Stringf( "a%u", index ) ) );
     gurobi.addLeqConstraint( terms, 0 );
 }
+
+void MILPEncoder::encodeMaxConstraint( GurobiWrapper &gurobi, MaxConstraint *max )
+{
+
+    if ( !max->isActive() )
+        return;
+
+    // y = max(x_1, x_2, ... , x_m)
+    unsigned y = max->getF();
+
+    // xs = [x_1, x_2, ... , x_m]
+    List<unsigned> xs = max->getParticipatingVariables();
+    unsigned m = xs.size();
+
+    // upper bounds of each x_i
+    double* ubs = new double[m];
+
+    // terms for Gurobi
+    List<GurobiWrapper::Term> terms;
+
+    unsigned i = 0;
+    for ( const auto &x : xs ) {
+        // add binary variable
+        gurobi.addVariable( Stringf( "a%u", x ),
+                            0,
+                            1,
+                            GurobiWrapper::BINARY );
+
+        terms.append( GurobiWrapper::Term( 1, Stringf( "a%u", x ) ) );
+        ubs[i] = _tableau.getUpperBound( x );
+        i++;
+    }
+
+    // add constraint: a_1 + a_2 + ... + a_m = 1
+    gurobi.addEqConstraint( terms, 1 );
+
+    terms.clear();
+
+    i = 0;
+    for ( const auto &x : xs ) {
+        // add constraint: y <= x_i + (1 - a_i) * (umax - l)
+        double umax = getUmax( ubs, i, m );
```

Comment:
if we store the <upper bound, variable> pair in a priority queue, we should be able to compute each Umax in O(log n) time. Currently it is computed in linear time, which might create overhead as the network size increases.

Reply:
Your idea may work, but I do not think it makes sense to me at the moment.
I think your idea needs to generate priority queue per [this loop](https://github.com/NeuralNetworkVerification/Marabou/blob/35fd279a3d018dd1b1791b630fc762d2e08c7299/src/engine/MILPEncoder.cpp#L171-L189) before calling _getUmax_ since the priorities are different per _i_.
However it requires log(n) to push elements into the priority queue for the each loop anyway.

I tried to decrease the computation cost by sharing upper bound array for every _i_.

If my understanding is not correct, please let me know.

What do you think?

Reply:
I did.